lib/manager/pip-compile/artifacts.ts

```diff
@@ -0,0 +1,102 @@
+import { quote as pipCompile } from 'shlex';
+import { TEMPORARY_ERROR } from '../../constants/error-messages';
+import { logger } from '../../logger';
+import { ExecOptions, exec } from '../../util/exec';
+import { deleteLocalFile, readLocalFile, writeLocalFile } from '../../util/fs';
+import { getRepoStatus } from '../../util/git';
+import type {
+  UpdateArtifact,
+  UpdateArtifactsConfig,
+  UpdateArtifactsResult,
+} from '../types';
+
+function getPythonConstraint(
+  config: UpdateArtifactsConfig
+): string | undefined | null {
+  const { constraints = {} } = config;
+  const { python } = constraints;
+
+  if (python) {
+    logger.debug('Using python constraint from config');
+    return python;
+  }
+
+  return undefined;
+}
+
+function getPipToolsConstraint(config: UpdateArtifactsConfig): string | null {
+  const { constraints = {} } = config;
+  const { pipTools } = constraints;
+
+  if (pipTools) {
+    logger.debug('Using pipTools constraint from config');
+    return pipTools;
+  }
+
+  return '';
+}
+
+export async function updateArtifacts({
+  packageFileName: inputFileName,
+  newPackageFileContent: newInputContent,
+  config,
+}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
+  const outputFileName = inputFileName.replace(/(\.in)?$/, '.txt');
```

Comment:
Because `fileMatch` allows arbitrary names, it might match non-`.in` files. And if we do search/replace `.in`->`.txt` then it would be chaos because it would return the same file name. This just ensures that we get a different file name.

The alternative is to enforce `.in` extensions regardless of `fileMatch` - which do you prefer?